src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -38,7 +48,132 @@ public static IEnumerable<ApplicationPart> DiscoverAssemblyParts(string entryPoi
+            var entryAssemblyAdditionalReferences = new List<AdditionalReference>();
+            foreach (var (assembly, metadata) in additionalReferences)
+            {
+                if (metadata.Length > 0)
+                {
+                    foreach (var metadataAttribute in metadata)
+                    {
+                        var fileName = Path.GetFileName(metadataAttribute.Value.Substring(0, metadataAttribute.Value.IndexOf(",")));
+                        var filePath = Path.Combine(Path.GetDirectoryName(assembly.Location), fileName);
+                        var additionalReference = new AdditionalReference
+                        {
+                            File = filePath,
+                            IncludeByDefault = metadataAttribute.Value.Substring(metadataAttribute.Value.IndexOf(",") + 1).Equals("true")
+                        };
+
+                        additionalAssemblies.Add(additionalReference);
+                        if (assembly.Equals(entryAssembly))
+                        {
+                            entryAssemblyAdditionalReferences.Add(additionalReference);
+                        }
+                    }
+                }
+                else
+                {
+                    // Fallback to loading the views like in previous versions if the additional reference metadata attribute is not present.
+                    var viewsAssemblies = GetViewAssemblies(assembly);
+                    foreach (var viewsAssembly in viewsAssemblies)
+                    {
+                        if (viewsAssembly != null)
+                        {
+                            var additionalReference = new AdditionalReference
+                            {
+                                File = null,
+                                Assembly = viewsAssembly,
+                                IncludeByDefault = true
+                            };
+
+                            additionalAssemblies.Add(additionalReference);
+                            if (entryAssembly.Equals(assembly))
+                            {
+                                entryAssemblyAdditionalReferences.Add(additionalReference);
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Load all the additional references.
+            for (var i = 0; i < additionalAssemblies.Count; i++)
+            {
+                var reference = additionalAssemblies[i];
+                if (reference.Assembly == null)
+                {
+                    var assembly = Assembly.LoadFile(reference.File);
+                    reference.Assembly = assembly;
+                }
+            }
+
+            // Discard any additional reference that was added by default through the dependency context.
+            var result = candidateAssemblies
+                .Where(ca => !additionalAssemblies.Any(ara => ara.Assembly.FullName.Equals(ca.FullName)))
```

Comment:
should additional assemblies be a dictionary or a set?

Reply:
This is an O(n^2) and it doesn't need to be.

Reply:
Are you concerned about the perfz? We discussed something similar an @eilon time ago when we did this feature for controllersfeatureprovider and at the time I showed that it didn't really matter unless you had like 10k assemblies I believe, but if you insist I'll fix it.